rootfs/etc/nginx/lua/util.lua

```diff
@@ -24,15 +26,69 @@ function _M.get_nodes(endpoints)
   return nodes
 end
 
--- given an Nginx variable i.e $request_uri
--- it returns value of ngx.var[request_uri]
-function _M.lua_ngx_var(ngx_var)
-  local var_name = string_sub(ngx_var, 2)
-  if var_name:match("^%d+$") then
-    var_name = tonumber(var_name)
+-- parse the compound variables, then call generate_var_value function
+-- to parse into a string value.
+function _M.parse_complex_value(complex_value)
+    local reg = [[ (\\\$[0-9a-zA-Z_]+) | ]]     -- \$var
+            .. [[ \$\{([0-9a-zA-Z_]+)\} | ]]    -- ${var}
+            .. [[ \$([0-9a-zA-Z_]+) | ]]        -- $var
+            .. [[ (\$|[^$\\]+) ]]               -- $ or text value
+    local iterator, err = re_gmatch(complex_value, reg, "jiox")
+    if not iterator then
+        return nil, err
+    end
+
+    local v
+    local t = {}
+    while true do
+        v, err = iterator()
+        if err then
+            return nil, err
+        end
+
+        if not v then
+            break
+        end
+
+        table.insert(t, v)
+    end
+
+    return t
+end
+
+-- Parse the return value of function parse_complex_value
+-- into a string value
+function _M.generate_var_value(data)
+  if data == nil then
+    return ""
   end
 
-  return ngx.var[var_name]
+  local t = {}
+  for _, value in ipairs(data) do
+    local var_name = value[2] or value[3]
+    if var_name then
+      if var_name:match("^%d+$") then
+        var_name = tonumber(var_name)
+      end
+      table.insert(t, ngx.var[var_name])
+    else
+      table.insert(t, value[1] or value[4])
+    end
+  end
+
+  return table.concat(t, "")
+end
+
+
+-- given the compound Nginx variables i.e {"$request_uri"} or {"$request_uri$host$scheme"}
+-- or {"${request_uri}-text-value"}, it returns the parsed value.
+function _M.lua_ngx_var(data)
```

Comment:
Can we delete this now? Does not seem to be used anymore.

Reply:
OK. I move function `lua_ngx_var` to `util_test.lua` file.

Reply:
What I meant this PR replaced `lua_ngx_var` with something else. Don't you think we should remove all references of `lua_ngx_var` (including tests), and add tests for the two new utility functions you introduced?

Reply:
Hi @ElvinEfendi The name of `lua_ngx_var` function is preserved but it contains two utility functions introduced this time.  Function `parse_complex_value` and `generate_var_value` need to be used together (split into two functions because of performance considerations), So they are not individually written unit tests.

Reply:
> split into two functions because of performance considerations

This indicates me they should be individually unit tested. But I don't feel strongly if you don't add individual unit tests for them in this PR. But IMHO we should definitely delete all references to `lua_ngx_var` as it is promoting less performant solution.

--

the integration test for them is already being done as part of chash balancer test.

Reply:
Hi @ElvinEfendi  What do you think of the latest commit?